File: src/core/EventBus.js
/**
 * SCAR — THE LAST CHOICE
 * EventBus.js — Central pub/sub event system
 * Author: Sirish (Lead/Integration)
 *
 * All game systems communicate through this bus.
 * No direct coupling between systems.
 */

export const EVENTS = {
  // Game lifecycle
  GAME_STARTED: 'GAME_STARTED',
  GAME_PAUSED: 'GAME_PAUSED',
  GAME_RESUMED: 'GAME_RESUMED',
  GAME_COMPLETED: 'GAME_COMPLETED',

  // Story progression
  INTRO_COMPLETE: 'INTRO_COMPLETE',
  ATTACK_STARTED: 'ATTACK_STARTED',
  ATTACK_COMPLETE: 'ATTACK_COMPLETE',
  SCAR_RECEIVED: 'SCAR_RECEIVED',

  // Level flow
  LEVEL_STARTED: 'LEVEL_STARTED',
  LEVEL_COMPLETED: 'LEVEL_COMPLETED',
  MISSION_STARTED: 'MISSION_STARTED',
  MISSION_OBJECTIVE_UPDATED: 'MISSION_OBJECTIVE_UPDATED',
  MISSION_COMPLETED: 'MISSION_COMPLETED',
  MISSION_FAILED: 'MISSION_FAILED',

  // Power system
  POWER_AWAKENED: 'POWER_AWAKENED',
  POWER_PATH_CHANGED: 'POWER_PATH_CHANGED',
  POWER_LEVEL_UP: 'POWER_LEVEL_UP',

  // Choice system
  CHOICE_PRESENTED: 'CHOICE_PRESENTED',
  CHOICE_MADE: 'CHOICE_MADE',

  // Hero / antagonist
  HERO_DETECTED_PLAYER: 'HERO_DETECTED_PLAYER',
  HERO_ENCOUNTER: 'HERO_ENCOUNTER',
  HERO_RELATIONSHIP_CHANGED: 'HERO_RELATIONSHIP_CHANGED',

  // Combat
  PLAYER_DAMAGED: 'PLAYER_DAMAGED',
  PLAYER_HEALED: 'PLAYER_HEALED',
  ENEMY_DEFEATED: 'ENEMY_DEFEATED',

  // Ending
  FINAL_BATTLE_STARTED: 'FINAL_BATTLE_STARTED',
  FINAL_CHOICE_MADE: 'FINAL_CHOICE_MADE',
  ENDING_TRIGGERED: 'ENDING_TRIGGERED',

  // Score / Cloud
  SCORE_CALCULATED: 'SCORE_CALCULATED',
  SCORE_SUBMITTED: 'SCORE_SUBMITTED',

  // UI (consumed by Ashwidha's system)
  DIALOGUE_START: 'DIALOGUE_START',
  DIALOGUE_COMPLETE: 'DIALOGUE_COMPLETE',
  HUD_UPDATE: 'HUD_UPDATE',
  CINEMATIC_START: 'CINEMATIC_START',
  CINEMATIC_COMPLETE: 'CINEMATIC_COMPLETE',
};

class EventBus {
  constructor() {
    /** @type {Map<string, Set<Function>>} */
    this._listeners = new Map();
    this._history = []; // for debugging + QA
  }

  /**
   * Subscribe to an event.
   * @param {string} event - Event name from EVENTS
   * @param {Function} callback - Handler function
   * @returns {Function} Unsubscribe function
   */
  on(event, callback) {
    if (!this._listeners.has(event)) {
      this._listeners.set(event, new Set());
    }
    this._listeners.get(event).add(callback);

    // Return unsubscribe fn
    return () => this.off(event, callback);
  }

  /**
   * Subscribe once — auto-removes after first trigger.
   */
  once(event, callback) {
    const wrapper = (data) => {
      callback(data);
      this.off(event, wrapper);
    };
    return this.on(event, wrapper);
  }

  /**
   * Unsubscribe from an event.
   */
  off(event, callback) {
    if (this._listeners.has(event)) {
      this._listeners.get(event).delete(callback);
    }
  }

  /**
   * Emit an event with optional payload.
   * @param {string} event
   * @param {*} data
   */
  emit(event, data = null) {
    const entry = { event, data, timestamp: Date.now() };
    this._history.push(entry);

    if (this._listeners.has(event)) {
      this._listeners.get(event).forEach(cb => {
        try {
          cb(data);
        } catch (err) {
          console.error(`[EventBus] Error in handler for "${event}":`, err);
        }
      });
    }
  }

  /**
   * Get event history (for QA/debugging).
   */
  getHistory() {
    return [...this._history];
  }

  /**
   * Clear all listeners (call on game reset).
   */
  clear() {
    this._listeners.clear();
    this._history = [];
  }
}

// Singleton — one bus for the entire game
export const eventBus = new EventBus();

File: Assets/Scripts/Core/EventBus.cs
using System;
using System.Collections.Generic;

namespace Scar.Core
{
    public static class GameEvents
    {
        public const string GAME_STARTED = "GAME_STARTED";
        public const string PLAYER_MOVED = "PLAYER_MOVED";
        public const string PLAYER_DAMAGED = "PLAYER_DAMAGED";
        public const string COMBAT_STARTED = "COMBAT_STARTED";
        public const string COMBAT_ENDED = "COMBAT_ENDED";
        public const string ENEMY_DEFEATED = "ENEMY_DEFEATED";
        public const string SCAR_RECEIVED = "SCAR_RECEIVED";
        public const string POWER_AWAKENED = "POWER_AWAKENED";
        public const string CHOICE_MADE = "CHOICE_MADE";
        public const string POWER_PATH_CHANGED = "POWER_PATH_CHANGED";
        public const string HERO_DETECTED_PLAYER = "HERO_DETECTED_PLAYER";
        public const string HERO_ENCOUNTER = "HERO_ENCOUNTER";
        public const string FINAL_BATTLE_STARTED = "FINAL_BATTLE_STARTED";
        public const string BOSS_DEFEATED = "BOSS_DEFEATED";
        public const string FINAL_CHOICE_MADE = "FINAL_CHOICE_MADE";
        public const string ENDING_TRIGGERED = "ENDING_TRIGGERED";
        public const string GAME_OVER = "GAME_OVER";
    }

    public class EventBus
    {
        private static EventBus _instance;
        public static EventBus Instance => _instance ??= new EventBus();

        private readonly Dictionary<string, List<Action<object>>> _listeners = new();

        public void Subscribe(string eventName, Action<object> callback)
        {
            if (!_listeners.ContainsKey(eventName))
            {
                _listeners[eventName] = new List<Action<object>>();
            }
            _listeners[eventName].Add(callback);
        }

        public void Unsubscribe(string eventName, Action<object> callback)
        {
            if (_listeners.ContainsKey(eventName))
            {
                _listeners[eventName].Remove(callback);
            }
        }

        public void Publish(string eventName, object payload = null)
        {
            if (_listeners.TryGetValue(eventName, out var callbacks))
            {
                var copy = new List<Action<object>>(callbacks);
                foreach (var callback in copy)
                {
                    callback?.Invoke(payload);
                }
            }
        }

        public void Reset()
        {
            _listeners.Clear();
        }
    }
}
